README

   =================================================================
   =  primesieve - optimized sieve of Eratosthenes implementation  =
   =                                                               =
   =               http://primesieve.googlecode.com                =
   =================================================================

Contents
========

  1. About primesieve
  2. Algorithm complexity
  3. Requirements
  4. What's in this source archive
  5. Sieve of Eratosthenes implementation
  6. Motivation
  7. Further reading
  8. Reporting bugs
  9. References

1. About primesieve
===================

  primesieve is a free software program and C++ library that generates
  prime numbers and prime k-tuplets (twins, triplets, ...) < 2^64
  using a highly optimized implementation of the sieve of
  Eratosthenes.

  Homepage: http://primesieve.googlecode.com

2. Algorithm complexity
=======================

  primesieve uses the segmented sieve of Eratosthenes with wheel
  factorization, this algorithm has a complexity of O(n log log n)
  operations and uses O(n^0.5) space.

  primesieve's memory requirement per thread is about:
  pi(n^0.5) * 8 bytes + 400 kilobytes

3. Requirements
===============

  primesieve is written in C++03, it compiles with every standard
  compliant C++ compiler and runs on both little-endian (x86, x86-64)
  and big-endian (PowerPC, SPARC, ...) CPUs. The parallelization
  (optional) is implemented using OpenMP 2.0 or later.

  primesieve has been optimized for CPUs with 15 or more integer
  registers, so if you have a x86-64 CPU make sure to use a 64-bit
  version of primesieve (i386 has only 8 registers).

  For compilation instructions please refer to INSTALL.

4. What's in this source archive
================================

  ./doc - Documentation directory
  -------------------------------
  LIBPRIMESIEVE   Explains how to build primesieve as a C++ library
                  for use in your own program.
  EXAMPLES        Contains source code examples that explain how to
                  generate primes using libprimesieve.
  API             List of all public member functions of the
                  PrimeSieve and ParallelPrimeSieve C++ classes.
  FAQ             Frequently asked questions.

  ./src - Source code directory
  -----------------------------
  src/soe         Contains the sieve of Eratosthenes implementation.
  src/apps        Contains the primesieve console and gui
                  applications.
  src/test        Contains unit tests for the sieve of Eratosthenes
                  implementation (run `make test').

5. Sieve of Eratosthenes implementation
=======================================

  General information
  -------------------
  The src/soe directory contains the sieve of Eratosthenes source
  code. Files that start with a capital letter contain a class of the
  same name, files in lower case contain inline functions and
  constant definitions. If you are new to prime number sieving I
  recommend reading some of the papers from section 7.

  The doc/EXAMPLES file contains C++ source code examples that explain
  how to use PrimeSieve objects. The doc/API file contains the
  complete API of the PrimeSieve and ParallelPrimeSieve C++ classes.

  Quick start
  -----------
  The segmented sieve of Eratosthenes is in SieveOfEratosthenes.cpp.
  Multiples are crossed off in EratSmall.cpp, EratMedium.cpp and
  EratBig.cpp. Prime numbers are called back and printed in
  PrimeFinder.cpp using a macro from SieveOfEratosthenes-GENERATE.h.

  C++ classes
  -----------
  * ParallelPrimeSieve (derived from PrimeSieve)
    ParallelPrimeSieve provides an easy API for multi-threaded prime
    sieving using OpenMP, it scales well even with 100 CPU cores! In
    ParallelPrimeSieve each OpenMP thread sieves primes using a thread
    private PrimeSieve object. The doc/EXAMPLES file explains
    ParallelPrimeSieve in more detail.

  * PrimeSieve
    PrimeSieve objects provide an easy API for prime sieving,
    PrimeSieve::sieve(start, stop) sieves the primes within the
    interval [start, stop]. The classes below are all used directly or
    indirectly for sieving.

  * SieveOfEratosthenes (abstract class)
    Implementation of the segmented sieve of Eratosthenes using a bit
    array with 30 numbers per byte, each byte of the sieve array holds
    the 8 offsets k = { 7, 11, 13, 17, 19, 23, 29, 31 }.
    Its main methods are addSievingPrime(uint_t) which must be called
    consecutively for all primes <= sqrt(n) in order to sieve the
    primes up to n and sieve(). SieveOfEratosthenes uses the
    EratSmall, EratMedium and EratBig classes to cross-off multiples.

  * PrimeFinder (derived from SieveOfEratosthenes)
    Main SieveOfEratosthenes class: PrimeFinder is used to callback,
    print and count the primes and prime k-tuplets within the interval
    [start_, stop_] of the associated PrimeSieve object.

  * PrimeGenerator (derived from SieveOfEratosthenes)
    Generates the primes up to sqrt(n) needed for sieving by the
    PrimeFinder class.

  * PreSieve
    PreSieve objects are used to pre-sieve multiples of small primes
    e.g. <= 19 to speed up the sieve of Eratosthenes.

  * WheelFactorization (abstract class)
    Wheel factorization is used to skip multiples of small primes e.g.
    <= 7 to speed up the sieve of Eratosthenes. The abstract
    WheelFactorization class is used to initialize sieving primes i.e.
    addSievingPrime() calculates the first multiple >= start of each
    sieving prime and the position within the SieveOfEratosthenes
    array of that multiple (multipleIndex). The unsetBit() method is
    used to cross-off a multiple (unset a bit) and to calculate the
    sieving prime's next multiple.

  * EratSmall (derived from WheelFactorization)
    Segmented sieve of Eratosthenes algorithm with a hard-coded
    modulo 30 wheel that skips multiples of 2, 3 and 5. This algorithm
    is optimized for small sieving primes that have many multiples in
    each segment. EratSmall is a further optimized implementation of
    Achim Flammenkamp's algorithm.[9]

  * EratMedium (derived from WheelFactorization)
    Segmented sieve of Eratosthenes algorithm with a fixed modulo 210
    wheel that skips multiples of 2, 3, 5 and 7. The wheel is
    implemented using a precomputed lookup table (wheel210 array from
    WheelFactorization.cpp). This algorithm is optimized for medium
    sieving primes with a few multiples per segment.

  * EratBig (derived from WheelFactorization)
    Segmented sieve of Eratosthenes algorithm with Tomás Oliveira's
    improvement for big sieving primes [11] and a fixed modulo 210
    wheel that skips multiples of 2, 3, 5 and 7. The wheel is
    implemented using a precomputed lookup table (wheel210 array from
    WheelFactorization.cpp). EratBig is optimized for big sieving
    primes that have less than one multiple per segment.

  Miscellaneous
  -------------
  * config.h
    Contains compile time constants that set the size of various
    arrays and limits within primesieve. You can set these constants
    according to your CPU type to get the best performance.

6. Motivation
=============

  I (the author, <kim.walisch@gmail.com>) got my first PC in 1997, it
  had an Intel 200 mmx CPU and 32 megabytes of memory. I was 14 years
  old then and bought the book "C++ for Kids" because I wanted to
  program a computer game :-) Besides a simple graphic game the book
  also contained a trial division prime checker which I found amazing,
  one entered a positive integer and the PC started calculating, after
  about a minute the program printed "is prime" or "is not prime". I
  soon started to invent new faster algorithms ...

7. Further reading
==================

  Jonathan Sorenson's papers provide a good introduction to prime
  number sieves, the sieve of Eratosthenes and segmentation are
  explained in [5], wheel factorization is covered in [6].
  Tomás Oliveira e Silva on his website [11] explains his improvement
  to the segmented sieve of Eratosthenes for big sieving primes.
  A. Járai and E. Vatai's paper "Cache optimized linear sieve" [12]
  explains many of the optimizations used in primesieve.
  Joerg Richstein in [10] (German language) describes how to pre-sieve
  multiples of small primes to speed up the sieve of Eratosthenes.

8. Reporting bugs
=================

  To report a bug or give feedback please send an email to the author
  Kim Walisch <kim.walisch@gmail.com>.

9. References
=============

  [1] Richard C. Singleton, "An efficient prime number generator",
      Communications of the ACM 12, 563-564, 1969.
  [2] R. P. Brent, "The first occurrence of large gaps between
      successive primes", Mathematics of Computation, 27:959-963,
      1973.
  [3] C. Bays and R. Hudson, "The segmented sieve of Eratosthenes and
      primes in arithmetic progressions to 10^12", BIT 17:121 127,
      1977.
  [4] Paul Pritchard, "Fast compact prime number sieves (among
      others)", Journal of Algorithms 4 (1983), 332-344.
  [5] J. Sorenson, "An Introduction To Prime Number Sieves", Computer
      Science Technical Report Vol. 909, January 1990.
      http://research.cs.wisc.edu/techreports/1990/TR909.pdf
  [6] J. Sorenson, "An analysis of two prime number sieves", Computer
      Science Technical Report Vol. 1028, June 1991.
      http://research.cs.wisc.edu/techreports/1991/TR1028.pdf
  [7] J. Sorenson and I. Parberry, "Two Fast Parallel Prime Number
      Sieves", Information and Computation, Vol. 114, No. 1, 1994.
      http://larc.unt.edu/ian/pubs/sieve.pdf
  [8] J. Sorenson, "Trading Time for Space in Prime Number Sieves",
      Lecture Notes in Computer Science, Vol. 1423 (1998), 179-194.
  [9] Achim Flammenkamp, "The Art of Prime Sieving", 1998.
      http://wwwhomes.uni-bielefeld.de/achim/prime_sieve.html
 [10] Jörg Richstein, "Segmentierung und Optimierung von Algorithmen
      zu Problemen aus der Zahlentheorie", Gießen, Univ., Diss., 1999.
 [11] Tomás Oliveira e Silva, "Fast implementation of the segmented
      sieve of Eratosthenes", 2002.
      http://www.ieeta.pt/~tos/software/prime_sieve.html
 [12] A. Járai and E. Vatai, "Cache optimized linear sieve", Acta
      Univ. Sapientiae, Informatica, 3, 2 (2011) 205-223.
      http://www.acta.sapientia.ro/acta-info/C3-2/info32-5.pdf


CHANGES
Changes in version 4.2, 03/03/2012
==================================

1. libprimesieve generates (callback) primes up to 5% faster on
   little-endian CPUs due to improved endiansafe_cast.h.
2. The Makefile now supports the Solaris OS.
3. Fixed a bug in the primesieve GUI code (read BUGS).
4. Updated ExpressionParser.h (src/apps/*) to version 2.5.
5. Added nth prime example program (./examples).
6. Lots of refactoring to make the code easier to understand.

Changes in version 4.1, 12/01/2012
==================================

1. New PrimeSieveCallback interface class that simplifies prime
   number generation with classes, see doc/USAGE_EXAMPLES.
2. The primesieve console (terminal) application now supports
   GNU-style long options e.g. --size=256.
3. Fixed a bug (for sizeof(int) > 4) in the bit population count
   algorithm, read BUGS file.
4. Updated files: README, INSTALL, doc/USAGE_EXAMPLES.
5. The Makefile now uses the system's default C++ compiler instead
   of GNU g++ which makes it more portable.
6. The primesieve_error class has been moved into its own file
   src/soe/primesieve_error.h (previously PrimeSieve.h).

   API changes
   -----------

   The following two overly complex PrimeSieve methods have been
   removed:

   void PrimeSieve::generatePrimes(uint32_t start, uint32_t stop, void (*)(uint32_t, void*), void*);
   void PrimeSieve::generatePrimes(uint64_t start, uint64_t stop, void (*)(uint64_t, void*), void*);

   They are replaced by the following two new methods that greatly
   simplify prime number generation with classes, please refer
   to doc/USAGE_EXAMPLES for more information.

   void PrimeSieve::generatePrimes(uint32_t start, uint32_t stop, PrimeSieveCallback<uint32_t>*);
   void PrimeSieve::generatePrimes(uint32_t start, uint32_t stop, PrimeSieveCallback<uint64_t>*);

Changes in version 4.0, 18/10/2012
==================================

primesieve 4.0 is a major new release. A lot of work has been put into
documentation and improving primesieve as a library. The API of
primesieve 4.0 is backward incompatible with primesieve 3.*, the few
changes are documented further down.

 1. Added support for OpenMP 2.*, useful for compilers that do
    not support OpenMP >= 3.0, e.g. MSVC, Apple g++.
 2. New examples directory with 12 simple example programs.
 3. New doc/API file that lists the public member functions of the
    PrimeSieve and ParallelPrimeSieve C++ classes.
 4. The Makefile is now POSIX compatible, it works with any POSIX
    shell e.g. sh, bash, ash, ksh, zsh, ...
 5. The Makefile now supports MinGW (with MSYS) and Cygwin.
 6. Fixed a shared libprimesieve bug (read doc/BUGS).
 7. Up to 10 percent faster prime number generation due to new
    internal 64-bit getNextPrime() (previously 32-bit).
 8. New unsynchronized ParallelPrimeSieve::generatePrimes() method
    that calls back primes in parallel (read USAGE_EXAMPLES).
 9. Faster thread synchronization in ParallelPrimeSieve, replaced slow
    OpenMP critical directive with faster omp_test_lock().
10. Optimized prime k-tuplet (twin primes, ...) counting for
    out-of-order CPUs (src/soe/PrimeNumberFinder.cpp).
11. New primesieve_error() exception used for all exceptions within
    PrimeSieve and ParallelPrimeSieve (read USAGE_EXAMPLES).
12. New pre-sieve code (src/soe/PreSieve.cpp).
13. Ported the primesieve GUI application from Qt 4 to Qt 5.

    API changes
    -----------

    The PrimeSieve 3.* count methods:

    uint64_t getPrimeCount(uint64_t start, uint64_t stop);
    uint64_t getTwinCount (uint64_t start, uint64_t stop);
    ...

    Have been renamed to:

    uint64_t countPrimes     (uint64_t start, uint64_t stop);
    uint64_t countTwins      (uint64_t start, uint64_t stop);
    uint64_t countTriplets   (uint64_t start, uint64_t stop);
    uint64_t countQuadruplets(uint64_t start, uint64_t stop);
    uint64_t countQuintuplets(uint64_t start, uint64_t stop);
    uint64_t countSextuplets (uint64_t start, uint64_t stop);
    uint64_t countSeptuplets (uint64_t start, uint64_t stop);

Changes in version 3.8, 13/07/2012
==================================

 1. Improved OOP design of WheelFactorization.h.
 2. Minor speed up for big sieving primes ~2% (src/soe/EratBig.cpp),
    reduced the number of operations in the main sieving loop.
 3. Minor speed up for small sieving primes ~3% (src/soe/EraSmall.cpp),
    new inner sieving loop without instruction dependencies that uses
    only 12 asm instructions (previously 16).
 4. Improved OpenMP load balance in src/soe/ParallelPrimeSieve.cpp.
 5. Improved code readability of EratSmall.cpp, EratMedium.cpp,
    EratBig.cpp, SieveOfEratosthenes.cpp and others.
 6. The Makefile now automatically detects the CPU's L1 data cache
    size on Unix-like OSes (Linux, Mac OS X).
 7. Renamed ./docs to ./doc
 8. Revised README.txt, added 7. Motivation.
 9. Updated INSTALL, LIBPRIMESIEVE and USAGE_EXAMPLES.
10. Added version #defines to PrimeSieve.h, e.g. for this release:

    #define PRIMESIEVE_VERSION       "3.8"
    #define PRIMESIEVE_MAJOR_VERSION  3
    #define PRIMESIEVE_MINOR_VERSION  8
    #define PRIMESIEVE_YEAR           2012

Changes in version 3.7, 31/05/2012
==================================

 1. More aggressive inlining (*-inline.h), up to 10 percent faster
    prime number generation, up to 20 percent faster initialization.
 2. Reduced header file dependencies, libprimesieve now only depends on
    PrimeSieve.h or ParallelPrimeSieve.h (and PrimeSieve.h).
 3. Updated Makefiles.
 4. New template imath.h functions: isqrt(), ilog2(), isPowerOf2(), ...
    isqrt() has been rewritten using Newton's algorithm in order to
    avoid rounding errors of (int)sqrt((double)n) if n > 10^15.
 5. New internal SieveOfEratosthenes::getNextPrime(...) member function.
 6. Replaced old C-style comments with C++ comments.
 7. Updated ExpressionParser to version 2.2.
 8. uin32_t has been replaced by uint_t in src/soe/*.

soe/cpuid.h
/*
 * cpuid.h -- This file is part of primesieve
 *
 * Copyright (C) 2011 Kim Walisch, <kim.walisch@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
 
 /** 
 * @file cpuid.h 
 * @brief Contains a portable implementation of cpuid for x86 and x64
 *        CPUs and macros/functions that make use of compiler
 *        intrinsics or built-in functions.
 */

#ifndef CPUID_H
#define CPUID_H

#include <climits>
#include <cassert>

#define bit_POPCNT (1 << 23)

#if defined(_MSC_VER) && (defined(_WIN32) || defined(_WIN64)) || \
    defined(__INTEL_COMPILER) && (defined(_WIN32) || defined(_WIN64))
#  define MSC_X86_COMPATIBLE
#  include <intrin.h> /* __cpuid(), _BitScanForward() */
#elif defined(__GNUC__) && (defined(__i386__) || defined(__x86_64__)) || \
      defined(__SUNPRO_CC) && (defined(__i386) || defined(__x86_64))
#  define GCC_I386_COMPATIBLE
#  include "cpuidgcc460.h" /* gcc-4.6.0/gcc/config/i386/cpuid.h */
#endif

/**
 * @def POPCNT64(addr)
 * Count the number of 1 bits within the next 8 bytes of an array
 * using the SSE4.2 POPCNT instruction.
 * @pre Check if the CPU supports POPCNT with isPOPCNTSupported()
 * @pre addr must be aligned to an 8-byte boundary
 *
 * Successfully tested with:
 *
 *   Microsoft Visual Studio 2010 (WIN32, WIN64),
 *   GNU G++ 4.5 (x86, x64),
 *   Intel C++ Compiler 12.0 (x86, x64),
 *   Oracle Solaris Studio 12 (x64).
 */
#if defined(MSC_X86_COMPATIBLE) && defined(_WIN64)
#  include <nmmintrin.h> 
#  define POPCNT64(addr) static_cast<uint32_t> \
      (_mm_popcnt_u64(*reinterpret_cast<const uint64_t*> (addr)))
#elif defined(MSC_X86_COMPATIBLE) && defined(_WIN32)
#  include <nmmintrin.h>
#  define POPCNT64(addr) \
      (_mm_popcnt_u32(reinterpret_cast<const uint32_t*> (addr)[0]) + \
       _mm_popcnt_u32(reinterpret_cast<const uint32_t*> (addr)[1]))
#elif defined(__SUNPRO_CC) && defined(__x86_64)
#  include <nmmintrin.h>
#  define POPCNT64(addr) static_cast<uint32_t> \
      (_mm_popcnt_u64(*reinterpret_cast<const uint64_t*> (addr)))
#elif defined(__GNUC__) && defined(__x86_64__)
#  define POPCNT64(addr) static_cast<uint32_t> \
      (__builtin_popcountll(*reinterpret_cast<const uint64_t*> (addr)))
#elif defined(__GNUC__) && defined(__i386__) && INT_MAX >= 2147483647
#  define POPCNT64(addr) \
      (__builtin_popcount(reinterpret_cast<const uint32_t*> (addr)[0]) + \
       __builtin_popcount(reinterpret_cast<const uint32_t*> (addr)[1]))
#endif

/**
 * Portable implementation of cpuid for x86 and x64 CPUs.
 * @return 1 if the CPU supports the cpuid instruction.
 *         0 if the CPU does not support the cpuid instruction.
 *        -1 if the target architecture (or compiler) is not supported.
 *
 * Successfully tested with:
 *
 *   Microsoft Visual Studio 2010,
 *   GNU G++ 4.5,
 *   Apple G++ 4.2,
 *   LLVM 2.9,
 *   Intel C++ Compiler 12.0,
 *   AMD x86 Open64 Compiler Suite,
 *   Oracle Solaris Studio 12.
 */
inline int getCPUID(unsigned int __level, unsigned int *__eax,
    unsigned int *__ebx, unsigned int *__ecx, unsigned int *__edx) {
#if defined(MSC_X86_COMPATIBLE)
  int CPUInfo[4] = {*__eax, *__ebx, *__ecx, *__edx};
  __cpuid(CPUInfo, 0);
  // check if the CPU supports the cpuid instruction.
  if (CPUInfo[0] != 0) {
    __cpuid(CPUInfo, __level);
    *__eax = CPUInfo[0];
    *__ebx = CPUInfo[1];
    *__ecx = CPUInfo[2];
    *__edx = CPUInfo[3];
    return 1;
  }
  return 0;
#elif defined(GCC_I386_COMPATIBLE)
  // __get_cpuid() returns 1 if the CPU supports cpuid else 0
  return __get_cpuid(__level, __eax, __ebx, __ecx, __edx);
#else
  return -1;
#endif
}

/**
 * Check if the CPU supports the SSE4.2 POPCNT instruction using
 * cpuid.
 */
inline bool isPOPCNTSupported() {
  unsigned int info_type = 0x00000001;
  unsigned int ax, bx, cx, dx;
  if (getCPUID(info_type, &ax, &bx, &cx, &dx) == 1)
    return ((cx & bit_POPCNT) != 0);
  return false;
}

/**
 * Search the operand (v) for the least significant set bit (1 bit)
 * and return its position.
 * @pre v != 0
 */
inline uint32_t bitScanForward(uint32_t v) {
  assert(v != 0);
#if defined(MSC_X86_COMPATIBLE)
  unsigned long r;
  _BitScanForward(&r, static_cast<unsigned long> (v));
  return static_cast<uint32_t> (r);
#elif defined(__GNUC__) && INT_MAX >= 2147483647
  return static_cast<uint32_t> (__builtin_ctz(v));
#else
  // Count the consecutive zero bits (trailing) on the right with
  // multiply and lookup, code from "Bit Twiddling Hacks":
  // http://graphics.stanford.edu/~seander/bithacks.html#ZerosOnRightMultLookup
  static const uint32_t MultiplyDeBruijnBitPosition[32] = {
       0,  1, 28,  2, 29, 14, 24, 3,
      30, 22, 20, 15, 25, 17,  4, 8,
      31, 27, 13, 23, 21, 19, 16, 7,
      26, 12, 18,  6, 11,  5, 10, 9
  };
  const int32_t s = static_cast<int32_t> (v);
  return MultiplyDeBruijnBitPosition[((s & -s) * 0x077CB531U) >> 27];
#endif
}

#endif /* CPUID_H */
